src/Microsoft.Azure.WebJobs.Host/Bindings/QueueBindingProvider.cs

```diff
@@ -0,0 +1,262 @@
+        // Constructor is just for capturing instance fields used in func closures. 
+        private QueueBindingProvider(
+               IStorageAccountProvider accountProvider,
+               IContextGetter<IMessageEnqueuedWatcher> messageEnqueuedWatcherGetter)
+        {
+            _accountProvider = accountProvider;
+            _messageEnqueuedWatcherGetter = messageEnqueuedWatcherGetter;
+        }
+
+        public static IBindingProvider Build(
+            IStorageAccountProvider accountProvider,
+            IContextGetter<IMessageEnqueuedWatcher> messageEnqueuedWatcherGetter,
+            INameResolver nameResolver,
+            IConverterManager converterManager)
+        {
+            var closure = new QueueBindingProvider(accountProvider, messageEnqueuedWatcherGetter);
+            var bindingProvider = closure.New(nameResolver, converterManager);
+            return bindingProvider;
+        }
+
+        // Helper to allocate a binding provider for [Queue].
+        // private instance method since it's refering to lambdas that close over the instance fields. 
+        private IBindingProvider New(
+            INameResolver nameResolver,
+            IConverterManager converterManager)
+        {
+            // IStorageQueueMessage is the core testing interface 
+            converterManager.AddConverter<byte[], IStorageQueueMessage, QueueAttribute>(ConvertByteArrayToCloudQueueMessage);
+            converterManager.AddConverter<IStorageQueueMessage, byte[]>(ConvertCloudQueueMessageToByteArray);
+
+            converterManager.AddConverter<string, IStorageQueueMessage, QueueAttribute>(ConvertStringToCloudQueueMessage);
+            converterManager.AddConverter<IStorageQueueMessage, string>(ConvertCloudQueueMessageToString);
+
+            converterManager.AddConverter<CloudQueueMessage, IStorageQueueMessage>(ConvertToStorageQueueMessage);
+
+            var bindingFactory = new BindingFactory(nameResolver, converterManager);
+
+            var bindAsyncCollector = bindingFactory.BindToAsyncCollector<QueueAttribute, IStorageQueueMessage>(BuildFromQueueAttribute, ToWriteParameterDescriptorForCollector, CollectAttributeInfo);
+            var bindClient = bindingFactory.BindToExactAsyncType<QueueAttribute, IStorageQueue>(BuildClientFromQueueAttributeAsync, ToReadWriteParameterDescriptorForCollector, CollectAttributeInfo);
+            var bindSdkClient = bindingFactory.BindToExactAsyncType<QueueAttribute, CloudQueue>(BuildRealClientFromQueueAttributeAsync, ToReadWriteParameterDescriptorForCollector, CollectAttributeInfo);
+                        
+            var bindingProvider = new GenericCompositeBindingProvider<QueueAttribute>(
+                ValidateQueueAttribute, nameResolver, bindClient, bindSdkClient, bindAsyncCollector);
+
+            return bindingProvider;
+        }
+
+        // [Queue] has some legacy behavior where the storage account can be specified outside of the [Queue] attribute. 
```

Comment:
I'll rename "legacy" to "pre-existing".  My point was that new extensions shouldn't follow this pattern because it doesn't jive well with Functions.  So new extensions won't need to do this extra hook. 

---
In reply to: [64934730](https://github.com/Azure/azure-webjobs-sdk/pull/711#discussion_r64934730) [](ancestors = 64934730)